### PR TITLE
GH-41358: [R] Support join "na_matches" argument

### PR DESCRIPTION
**Pull request by nealrichardson**
### Rationale for this change

Noticed in #41350, I made #41358 to implement this in C++, but it turns out the option was there, just buried a bit.

### What changes are included in this PR?

`na_matches` is mapped through to the `key_cmp` field in `HashJoinNodeOptions`. Acero supports having a different value for this for each of the join keys, but dplyr does not, so I kept it constant for all key columns to match the dplyr behavior. 

### Are these changes tested?

Yes

### Are there any user-facing changes?

Yes
* GitHub Issue: #41358

### CODE REVIEW

**Comment by nealrichardson** [comment]
CI failure appears unrelated: https://github.com/apache/arrow/issues/41398

File: r/src/compute-exec.cpp [hunk 1]
```diff
@@ -411,10 +411,17 @@ std::shared_ptr<acero::ExecNode> ExecNode_Join(
...
+    bool na_matches) {
   std::vector<arrow::FieldRef> left_refs, right_refs, left_out_refs, right_out_refs;
+  std::vector<acero::JoinKeyCmp> key_cmps;
   for (auto&& name : left_keys) {
     left_refs.emplace_back(std::move(name));
+    // Populate key_cmps in this loop, one for each key
+    // Note that Acero supports having different values for each key, but dplyr
+    // only supports one value for all keys, so we're only going to support that
+    // for now.
+    key_cmps.emplace_back(na_matches ? acero::JoinKeyCmp::IS : acero::JoinKeyCmp::EQ);
```
Comment (jonkeane) [inline r/src/compute-exec.cpp:424]: Out of my curiosity to make sure I'm following this all: `acero::JoinKeyCmp::IS` ends up being used when `na_match == "na"`, and therefore `NA`s will match. `acero::JoinKeyCmp::EQ` ends up in all other cases (i.e. `na_match == "never"`) and that uses equality where `NA`s _do not_ match.

**Review by jonkeane — APPROVED** [review]
Thanks for this! I have one question, but it's really just to confirm that I'm reading this correctly.

File: r/src/compute-exec.cpp (hunk 1 above)
Comment (nealrichardson) [reply r/src/compute-exec.cpp:424]: Correct

Comment (jonkeane) [reply r/src/compute-exec.cpp:424]: So many ways of swapping and passing and evaluating booleans 😂

**Pull request merged**